einops/einops.py

```diff
@@ -624,3 +625,132 @@ def asnumpy(tensor) -> 'numpy.ndarray':
         `numpy.ndarray`, converted to numpy
     """
     return get_backend(tensor).to_numpy(tensor)
+
+
+@functools.lru_cache(256)
+def _compactify_pattern_for_einsum(pattern: str) -> str:
+    if "->" not in pattern:
+        # numpy allows this, so make sure users
+        # don't accidentally do something like this.
+        raise ValueError("Einsum pattern must contain '->'.")
+    lefts, right = pattern.split('->')
+    lefts = lefts.split(',')
+
+    lefts = [
+        ParsedExpression(left, allow_underscore=True, allow_duplicates=True)
```

Comment:
Looks like it will pass things like "b 1 h w->c 1" and assign "1" as one of axes.
Worth adding tests that parsing fails on the following cases:
"i 1 -> 1 i"
"i () -> i"
"i (j k) -> i j k"

Reply:
Good idea. Adding now.

Reply:
Done.